src/ussd_sassa_registration.js

```diff
@@ -67,109 +93,389 @@ go.app = (function () {
-                    ).context({
+          if (_.upperCase(content) != "NO") {
+            var match = content.match(/^(\d{6})(\d{4})(0|1)8\d$/);
+            var today = new moment(self.im.config.testing_today).startOf("day"),
+              dob;
+            var validLuhn = function (content) {
+              return (
+                content
+                  .split("")
+                  .reverse()
+                  .reduce(function (sum, digit, i) {
+                    return (
+                      sum +
+                      _.parseInt(
+                        i % 2 ? [0, 2, 4, 6, 8, 1, 3, 5, 7, 9][digit] : digit
+                      )
+                    );
+                  }, 0) %
+                  10 ==
+                0
+              );
+            };
+            if (
+              !match ||
+              !validLuhn(content) ||
+              !(dob = new moment(match[1], "YYMMDD")) ||
+              !dob.isValid() ||
+              !dob.isBetween(
+                today.clone().add(-130, "years"),
+                today.clone().add(-17, "years")
+              )
+            ) {
+              return $(error_msg);
+            } else {
+              return new JsonApi(self.im)
+                .post(
+                  self.im.config.sassa_api.url + "/api/v1/check_id_number",
+                  {
+                    data: {
                       id_number: content,
-                      status: response.data.status,
-                    });
+                    },
+                    headers: {
+                      Authorization: [
+                        "Token " + self.im.config.sassa_api.token,
+                      ],
+                      "User-Agent": ["Jsbox/SASSA-Registration-USSD"],
+                    },
                   }
-                },
-                function (e) {
-                  // Go to error state after 3 failed HTTP requests
-                  opts.http_error_count =
-                    _.get(opts, "http_error_count", 0) + 1;
-                  if (opts.http_error_count === 3) {
-                    self.im.log.error(e.message);
-                    return self.states.create("__error__", {
-                      return_state: name,
-                    });
+                )
+                .then(
+                  function (response) {
+                    if (!response.data.valid || response.data.underage) {
+                      return $(error_msg);
+                    } else if (response.data.existing) {
+                      return $(
+                        [
+                          "An application for {{ id_number }} has already been made. The status is: {{ status }}.",
+                          "To appeal, call 0800 002 9999",
+                        ].join("\n")
+                      ).context({
+                        id_number: content,
+                        status: response.data.status,
+                      });
+                    }
```

Comment:
We should probably also have some error handling for if the request fails. Maybe not right now, but at some point.